tests/integration_tests/functional/test_uffd.py

```diff
@@ -0,0 +1,238 @@
+
+    return proc
+
+
+def test_bad_socket_path(bin_cloner_path, test_microvm_with_api):
+    """
+    Test error scenario when socket path does not exist.
+
+    @type: negative
+    """
+    logger = logging.getLogger("uffd_bad_socket_path")
+
+    logger.info("Create snapshot")
+    snapshot = create_snapshot(bin_cloner_path)
+
+    logger.info("Load snapshot, mem %s", snapshot.mem)
+    vm = test_microvm_with_api
+    vm.spawn()
+    jailed_vmstate = vm.create_jailed_resource(snapshot.vmstate)
+
+    response = vm.snapshot.load(
+        mem_backend={
+            'type': SnapshotMemBackendType.UFFD,
+            'path': 'inexsistent'
+        },
+        snapshot_path=jailed_vmstate
+    )
+
+    assert vm.api_session.is_status_bad_request(response.status_code)
+    assert "Load microVM snapshot error: Cannot connect to UDS in order to " \
+           "send information on handling guest memory page-faults due to: " \
+           "No such file or directory (os error 2)" in response.text
+
+
+def test_unbinded_socket(bin_cloner_path, test_microvm_with_api):
+    """
+    Test error scenario when PF handler has not yet called bind on socket.
+
+    @type: negative
+    """
+    logger = logging.getLogger("uffd_unbinded_socket")
+
+    logger.info("Create snapshot")
+    snapshot = create_snapshot(bin_cloner_path)
+
+    logger.info("Load snapshot, mem %s", snapshot.mem)
+    vm = test_microvm_with_api
+    vm.spawn()
+    jailed_vmstate = vm.create_jailed_resource(snapshot.vmstate)
+
+    socket_path = os.path.join(vm.path, "firecracker-uffd.sock")
+    run_cmd("touch {}".format(socket_path))
+    jailed_sock_path = vm.create_jailed_resource(socket_path)
+
+    response = vm.snapshot.load(
+        mem_backend={
+            'type': SnapshotMemBackendType.UFFD,
+            'path': jailed_sock_path
+        },
+        snapshot_path=jailed_vmstate
+    )
+
+    assert vm.api_session.is_status_bad_request(response.status_code)
+    assert "Load microVM snapshot error: Cannot connect to UDS in order to" \
+           " send information on handling guest memory page-faults due to: " \
+           "Connection refused (os error 111)" in response.text
+
+
+def test_valid_handler(bin_cloner_path,
+                       test_microvm_with_api,
+                       uffd_handler_paths):
+    """
+    Test valid uffd handler scenario.
+
+    @type: functional
+    """
+    logger = logging.getLogger("uffd_unbinded_socket")
+
+    logger.info("Create snapshot")
+    snapshot = create_snapshot(bin_cloner_path)
+
+    logger.info("Load snapshot, mem %s", snapshot.mem)
+    vm_builder = MicrovmBuilder(bin_cloner_path)
+    vm = test_microvm_with_api
+    vm.spawn()
+
+    # Copy snapshot memory file into chroot of microVM.
+    jailed_mem = vm.create_jailed_resource(snapshot.mem)
+    # Copy the valid page fault binary into chroot of microVM.
+    jailed_handler = vm.create_jailed_resource(
+        uffd_handler_paths['valid_handler']
+    )
```

Comment:
could we also include this in `spawn_pf_handler`? Just so that it's less boilerplate code when adding a new test